src/merkle_tree/src/storage_proof.cairo

```diff
@@ -0,0 +1,160 @@
+) -> felt252 {
+    let contract_data = proof.contract_data;
+
+    let (contract_root_hash, storage_value) = traverse(
+        storage_address, contract_data.storage_proof
+    );
+
+    let contract_state_hash = pedersen(
+        pedersen(pedersen(contract_data.class_hash, contract_root_hash), contract_data.nonce),
+        contract_data.contract_state_hash_version
+    );
+
+    let (contracts_tree_root, expected_contract_state_hash) = traverse(
+        contract_address, proof.contract_proof
+    );
+
+    assert(expected_contract_state_hash == contract_state_hash, 'wrong contract_state_hash');
+
+    let state_commitment = poseidon_hash(
+        'STARKNET_STATE_V0', contracts_tree_root, proof.class_commitment
+    );
+
+    assert(expected_state_commitment == state_commitment, 'wrong state_commitment');
+
+    storage_value
+}
+
+fn traverse(expected_path: felt252, proof: Array<TrieNode>) -> (felt252, felt252) {
+    let mut path: felt252 = 0;
+    let mut remaining_depth: u8 = 251;
+
+    let mut nodes = proof.span();
+    let expected_path_u256: u256 = expected_path.into();
+
+    let leaf = *match nodes.pop_back().unwrap() {
+        TrieNode::Binary(_) => panic_with_felt252('invalid leaf type'),
+        TrieNode::Edge(edge) => edge
+    };
+
+    let mut expected_hash = node_hash(@TrieNode::Edge(leaf));
+    let value = leaf.child;
+    let mut path = leaf.path;
+    let mut path_length_pow2 = pow(2, leaf.length);
+
+    loop {
+        match nodes.pop_back() {
+            Option::Some(node) => {
+                match node {
+                    TrieNode::Binary(binary_node) => {
+                        if expected_path_u256 & path_length_pow2.into() > 0 {
+                            assert(expected_hash == *binary_node.right, 'invalid node hash');
+                            path += path_length_pow2;
+                        } else {
+                            assert(expected_hash == *binary_node.left, 'invalid node hash');
+                        };
+                        path_length_pow2 *= 2;
+                    },
+                    TrieNode::Edge(edge_node) => {
+                        assert(expected_hash == *edge_node.child, 'invalid node hash');
+                        path += *edge_node.path * path_length_pow2;
+                        path_length_pow2 *= pow(2, *edge_node.length);
+                    }
+                }
+                expected_hash = node_hash(node);
+            },
+            Option::None => { break; }
+        };
+    };
+    assert(expected_path == path, 'invalid proof path');
+    (expected_hash, value)
+}
+
+#[inline]
+fn node_hash(node: @TrieNode) -> felt252 {
+    match node {
+        TrieNode::Binary(binary_node) => { pedersen(*binary_node.left, *binary_node.right) },
+        TrieNode::Edge(edge_node) => {
+            pedersen(*edge_node.child, *edge_node.path) + (*edge_node.length).into()
+        }
+    }
+}
+
+fn pow(x: felt252, n: u8) -> felt252 {
```

Comment:
This fn exists in math, please re-use it 😃

Reply:
Actually since you are only using pow of 2 I'd hardcode them. It'll be way more efficient:
Are are all of 'em
https://github.com/gaetbout/starknet-felt-packing/blob/main/contracts/lib/pow2.cairo

Reply:
I thought about but it but I could not find a way to declare an const array. Is there a Cairo 1 syntax equivalent to dw?

Reply:
I prefer to keep it as is for now. Added TODO to turn it into a loopup table later.